pkg/image/registryclient/client.go

```diff
@@ -200,30 +215,80 @@ func (c *Context) Ping(ctx context.Context, registry *url.URL, insecure bool) (h
 }
 
 func (c *Context) Repository(ctx context.Context, registry *url.URL, repoName string, insecure bool) (distribution.Repository, error) {
-	named, err := reference.WithName(repoName)
+	retryRepo, err := c.repository(ctx, registry, repoName, insecure)
 	if err != nil {
 		return nil, err
 	}
+	return retryRepo.Repository, nil
+}
+
+func (c *Context) repository(ctx context.Context, registry *url.URL, repoName string, insecure bool) (*retryRepository, error) {
+	// populate retryRepository with alternative strategy
+	limiter := c.Limiter
+	if limiter == nil {
+		limiter = rate.NewLimiter(rate.Limit(5), 5)
+	}
+	retryRepo := NewLimitedRetryRepository(nil, c.Retries, limiter)
+	if c.AlternateRepositoryStrategy != nil {
+		retryRepo.alternateStrategy = c.AlternateRepositoryStrategy
+	}
+	named, err := reference.WithName(repoName)
+	if err != nil {
+		return retryRepo, err
+	}
 
 	rt, src, err := c.Ping(ctx, registry, insecure)
 	if err != nil {
-		return nil, err
+		return retryRepo, err
 	}
 
 	rt = c.repositoryTransport(rt, src, repoName)
 
 	repo, err := registryclient.NewRepository(named, src.String(), rt)
 	if err != nil {
-		return nil, err
+		return retryRepo, err
 	}
 	if !c.DisableDigestVerification {
 		repo = repositoryVerifier{Repository: repo}
 	}
-	limiter := c.Limiter
-	if limiter == nil {
-		limiter = rate.NewLimiter(rate.Limit(5), 5)
+	retryRepo.Repository = repo
+	return retryRepo, nil
+}
+
+// RepositoryWithAlternateReference returns a a valid repository and ref, or will gather alternate sources then return first repo, ref, or error
+func (c *Context) RepositoryWithAlternateReference(ctx context.Context, locator imagereference.DockerImageReference, insecure bool) (distribution.Repository, imagereference.DockerImageReference, error) {
+	repo, err := c.repository(ctx, locator.DockerClientDefaults().RegistryURL(), locator.RepositoryName(), insecure)
```

Comment:
This call is not desirable for the image registry, the registry should access mirrors first.

Reply:
I'll update but oc may prefer to try original image sources first, since the `oc adm release` commands are opt out for alternate image source lookup, they implicitly lookup icsp. If there is a strong feeling against oc trying mirrored sources first, then i'll (add a boolean? separate fn?)